### ISSUE TITLE
much smaller contig N50s, same data -- a change somewhere between release 1.0 and 1.1

### ISSUE THREAD

**Issue by JohnUrban** [marbl/canu#80]
I have been updating Canu as frequently as possible over the course of the past few months. Unfortunately, I have not kept specific notes on versions -- so I will have to speak in terms of before/after/between releases 1.0 and 1.1 (and some dates).

I have a PacBio dataset that gave NG50s of 2-5 Mb with Canu before, around, and after the release of 1.0.

This morning, using the most current version, I noticed the NG50s were around 60-800 kb. I believe this trend of considerably lower NG50s occurred at some point between release 1.0 and 1.1. I first noticed the smaller-than-expected continuity statistics with other data (call it "newdata") on **March 1st**. For a while I assumed it had to do with the quality of the newdata, but in the past week I wanted to see if it was actually an issue with Canu. When running Canu on the "control" PacBio dataset discussed above that was used for multiple assemblies from previous versions, the contig stats were worse than they were for the "newdata". So it does not seem like the "newdata" is the problem per se. 

Although bigger continuity statistics do not signify a better assembly, I think in this case they do. I have HGAP3 and Falcon assemblies with this same "control" data that are comparable to the older Canu assemblies in terms of contig stats (i.e. size stats in 2-5 Mb range, not 60-800kb). I have reasons to believe that those assemblies are of good quality (i.e. not just a lot of mis-joins).

Unfortunately, our compute cluster is down for maintenance until Friday and I currently cannot access it to obtain more details, try new things, etc. I just wanted to raise this possible red flag in case it can help identify an issue anyway. 

Perhaps one possibility is that the new way of reporting contigs, bubbles, and unassembled instead of just one asm file has affected the stats I am seeing. I currently try to only obtain contig stats from the contigs file (not bubbles or unassembled). However, there seem to be very few "bubbles" all less than 50kb and there seems to be a comparable number of contigs in the contigs file as there used to be int eh asm file. Having said that, the assembly size seems to be at least 20 Mb smaller...

Another possibility is that this may be related to issue #63 where using "low coverage" settings led to really small contig stats...

**Comment by skoren** [comment]
Yes, there is a bug in Canu v1.1 with how the overlaps are filtered before generating corrected reads. I updated the release notes/docs to reflect this. Have you tried running your new dataset with that suggestion to see if it is comparable to the older stats?

'There is a bug with filtering overlaps before passing them to falcon_sense for generating corrected reads on repetitive genomes. If you are assembling large/repetitive genomes (generally >500Mb), you must specify 'corMaxEvidenceErate=0.2' ('corMaxEvidenceErate=0.3' for low-coverage datasets) while the fix is tested.'

**Comment by JohnUrban** [comment]
I have not tried this with the control dataset, but I did try a few things with the "newdata".
1. default settings
2. corMaxEvidenceErate=0.3
3. corMaxEvidenceErate=0.3 corMhapSensitivity=fast
4. corMaxEvidenceErate=0.3 corMhapSensitivity=fast minReadLength=3000 minOverlapLength=3000 corOutCoverage=80 (to use all data)
5. minReadLength=3000 minOverlapLength=3000 corOutCoverage=80

All did better than default settings, but not much better, and not as well as the older Canu assemblies (or HGAP3/Falcon). The settings in 5 did best.

**Comment by skoren** [comment]
There is a pretty big difference between the filtering at 0.3 and 0.2, especially on a high-coverage repetitive genome. The main symptom of the bug is that there are fewer corrected reads output which seems consistent with your results that setting corOutCoverage higher improves the result.

**Comment by JohnUrban** [comment]
Alright - I will certainly launch some jobs as soon as I can used corMaxEvidenceErate=0.2 to give it a try.

**Comment by skoren** [comment]
If you pull the latest from head it should properly do the filtering.

**Comment by rafalwoycicki** [comment]
Hi Sergey,

CAn I ask whet was the nature of this error? I am running now the correction of a large number of data (150x to correct, using 218x in total) without specyfing corMaxEvidenceLenght, just corErate. I have 90Mbp not repetetive genome. Should I worry? Stopit and rerun?
Rafal

**Comment by skoren** [comment]
corErate? Do you mean corErrorRate? That won't change the set of overlaps being used to generate corrected reads, generally I don't think it's useful to change it from the default.

It was an interaction between MHAP and the update to falcon_sense we made in 1.1. MHAP was producing more overlaps than needed for correction but the previous falcon_sense masked this because it was always using a subset of reads to call the consensus. The updated falcon_sense did not have this feature and the extra overlaps produced by MHAP are causing worse consensus for corrected reads/more splitting and as a result fewer long reads for assembly on repetitive genomes. We did not see this on our test sets of D. melanogaster, C. elegans, or A. thaliana, which are all larger than your genome so I wouldn't expect it to be an issue. You should be able to see if it will be an issue if you get significantly less than 30X corrected reads or if the corrected reads average length is shorter than the average length of your uncorrected reads.

**Comment by rafalwoycicki** [comment]
Thank you Sergey! (Yes, sorry, I meant corErrorRate). Ok, I got it. Thanks.

**Comment by rafalwoycicki** [comment]
So actually I had the problem with significantly lower coverage of corrected reads in canu 1.1 comparing to canu 1.0 (3x lower), when using defaults, and also when changing overlaps lengths.
The total assembly length from canu 1.1 was about 20 mbp less than from canu 1.0, and the contigs were about 3x shorter. The good thing was that bigger portion of the assembled genome in canu1.1 was correctly assembled - I mean more contigs were reflecting the contings from the reference.
I will try to run tomorrow canu 1.1 after this fix, with the earlier settings, and will see what will be the effect - hope for the best !:-)

**Comment by rafalwoycicki** [comment]
Hi, so I run the correction still using now the updated version, and the read correction step takes now 2x more time than before - did not finished yet. I am using 60x for correction out of 105x (with 2000/4000 ovl/minlength, and the rest at defaults).

**Comment by skoren** [comment]
Nothing in the commits since 1.1 would have affected the runtime, unless it is generating significantly more corrected reads which would take longer to compute since there are more of them.

**Comment by rafalwoycicki** [comment]
Thanks.
Can say that after the correction step I have gotten 50x coverage, instead of 26x like before the change you have made. Hope that now the assembly also will go better. Will let you know tomorrow. Great!

**Comment by rafalwoycicki** [comment]
Ok, so I have gotten now simialr results which I got with canu. Thanks.

**Comment by JohnUrban** [comment]
May this will be of interest. The size stats outcome is correlated with coverage after correction as you suggested -- in addition to other tweaks.

Above I said there were 5 conditions:
1. default settings
2. corMaxEvidenceErate=0.3
3. corMaxEvidenceErate=0.3 corMhapSensitivity=fast
4. corMaxEvidenceErate=0.3 corMhapSensitivity=fast minReadLength=3000 minOverlapLength=3000 corOutCoverage=80 (to use all data)
5. minReadLength=3000 minOverlapLength=3000 corOutCoverage=80
….Then there were the settings you suggested in #63 
6. corMinCoverage=0
7. corMhapSensitivity=high  corMaxEvidenceErate=0.7

Looking at coverage before and after correction step and after trimming:
1. --do not have info but untigGuessXcov was 26   
2. 51.5 --> 26.44 -->  26.40   trimGuess=28  unitigGuess=28
3. 51.5 --> 23.73 --> 23.69   trimGuess=23 unitigGuess=23
4. 48.1 --> 22.45 --> 22.41     trimGuess=22 unitigGuess=22
5.  48.1 --> 25.23 --> 25.19     trimGuess=27 unitigGuess=27
6. 51.5 --> 40.10 --> 29.02   trimGuess=28 unitigGuess=27
7. 51.5 --> 11.89 --> 11.78     trimGuess=102  unitigGuess=110

5 did the best
1, 2, and 6 are somewhat in tie for 2nd
3 and 4 are close 3rds
7 is worst

Our compute cluster is now running again, so I can download the new code and see if it fixed the issues in this thread and in #63 in general.

**Comment by skoren** [comment]
@rafalwoycicki do you mean your most recent results are more similar to Canu 1.0? @JohnUrban if your tests with the latest code restore the N50, I will push a bug fix release. Are either of you able to share your dataset with us, our tests didn't catch this filtering bug even though I include a human genome in our testing so I'd like to include a test case which catches this issue in our regression testing.

**Comment by rafalwoycicki** [comment]
Hi Sergey, yes, I have gotten similar results to canu 1.0. I would be happy to share my dataset with you, maybe we would find some way to assemble it properly (just need to ask my boss).

**Comment by rafalwoycicki** [comment]
Hi Sergey, how can I share the dataset with you?

**Comment by JohnUrban** [comment]
Hi Sergey - I'd be happy to share my data. 

In other news - I began running the new code on 4 conditions: 
1. default
2. low-cov settings 
3. default but corOutCoverage set higher than total coverage
4. low-cov and corOutCoverage set higher than total coverage

Oddly for 3 and 4, both ended in the correction phase with the following message in canu.out:

>Don't panic, but a mostly harmless error occurred and canu failed.
canu failed with 'Failed to find any overlaps (0.000) or slices (2).

This was after mhap at the ovStoreBuild step.
Any idea what happened?

**Comment by skoren** [comment]
@rafalwoycicki you can upload to ftp://ftp.cbcb.umd.edu/incoming/sergek. You won't be able to list files but you'll have write permissions. Only I can read the files for security.

**Comment by skoren** [comment]
@JohnUrban for 3,4 I'm guessing something didn't run in 1-overlapper, any error outputs from the canu.*.out files or mhap.*.out?

**Comment by JohnUrban** [comment]
As far as I can tell there are no errors reported in the mhap.out files … but all that I checked end with something like:
Total matches found: **0**
Average number of matches per lookup: **0.0**
Average number of table elements processed per lookup: 435.7253594771242
Average number of table elements processed per match: **Infinity**
Average % of hashed sequences hit per lookup: 0.6772991124021626
Average % of hashed sequences hit that are matches: **0.0**
Average % of hashed sequences fully compared that are matches: **NaN**

**Comment by skoren** [comment]
What's in mhap.sh and precompute.sh/out?

**Comment by JohnUrban** [comment]
One difference in the mhap.sh command seems to be:
--min-store-length 36251 (when corcovout set high)
--min-store-length 2243 (when default)

**Comment by skoren** [comment]
Are you sure the genome size was set the same for both?

**Comment by JohnUrban** [comment]
No errors as far as I can tell in the precompute.out files.

Same thing I saw in mhap.sh is in precompute.sh
--min-store-length 36251 (when corcovout set high)
--min-store-length 2243 (when default)

Seems to be the only difference in the mhap command.

**Comment by JohnUrban** [comment]
Just double-checked genome size and it is definitely the same.
I just have 4 canu commands in a row that all use the same variable for genome size.

**Comment by JohnUrban** [comment]
I can say that this did not happen before downloading the latest code… but happened twice since (for default and lowcov settings only when corOutCoverage set high).

**Comment by skoren** [comment]
The min store length is computed as the shortest read need to get to corOutCoverage based on the genome size. The code there hasn't changed since release 1.1 so it shouldn't be any different than it was and I've run several recent runs using corOutCoverage between 40 and 200 (exceeding the available coverage). There should be output like:
```
 -- Computed seed length 16485 from desired output coverage 40 and genome size 3100000000
```
What does it say for your four runs?

**Comment by JohnUrban** [comment]
Well when I checked that it said:
Computed seed length 16485 from desired output coverage  and genome size 310000000

which made me realize I had a variable:
CORCOVOUT=80

but in the canu command was using:
$COROUTCOV

So I was feeding --corOutCoverage an empty variable.

Pardon me for not checking that first. All is in working order though - so thats good. :)

**Comment by skoren** [comment]
So you specified corOutCoverage=""" essentially? It should have checked for that, I'll open a separate issue.

**Comment by JohnUrban** [comment]
Yes - exactly. I agree - it would be good for it to fail upfront and complain about what it needs.

**Comment by JohnUrban** [comment]
Alright so I have one example so far of how 1.0, 1.1-old, and 1.1-new are doing in terms of size stats.
These are just default settings used.

**Metric                             Canu 1.0            Canu 1.1-old        Canu 1.1-new**
Number contigs:            1,857                    1,614                     1,042
Assembly size:               308,490,429        284,094,258           292,356,748
Max contig size:             16,066,196          9,569,671               10,698,503
Min contig size:               536                     8,787                      9,968
Mean contig size:           166,123               176,019                  280,573
Median contig size:        15,220                 59,263                    70,542
Contig N50:	               2,359,081            555,811                 1,588,191
E size (G=asm_size):     3,969,506            1,766,026               2,778,616
Contig N50 (G=210m):   4,670,699            1,092,791               3,335,559
Contig N50 (G=350m):   1,509,919            312,081                  970,685
E size (G=210m):            5,831,213           2,389,132               3,868,319
E size (G=350m):            3,498,727           1,433,479               2,320,991

**Comment by JohnUrban** [comment]
Overall, as you can see above, the latest code brings the size stats closer to what they were with 1.0, but still a little lower. The total asm size is still smaller as well, and there are fewer contigs. There are only a handful of contigs in bubbles.fasta that sum to 100-200kb or so. Interestingly, if I take the sizes of the contigs out of the 5-consensus/consensus.out files, there are a couple hundred more contigs and ~5 Mb extra. These seem to be relegated to unassembled.fasta though. They range in length from 5.6-30.9 kb, contain 2-68 reads, and have covStats values of 1.24-69.81. I assume based on previous issue conversations I read that these are being relegated to the unassembled file possibly because 1 read makes up the entire length….(?) Still I wonder if they contain sequences worth keeping based on their coverage (mostly in line with expected cov).

One observation is that the min contig sizes are always bigger lately -- my guess is many of the contigs smaller than the new mins are ones relegated to unassembled.

I don't have comparisons for Canu 1.0 and 1.1-old, but for 1.1-new if I set the 2nd value in -repeatdetect to 2X or 3X the guessed X coverage, I get:

Metric | 2X | 3X
------- | ---- | -----
Number contigs:  | 1,021 | 1,010
Assembly size:  | 292,148,143 | 292,075,595
Max contig size:  | 10,709,908 | 16,410,386
Min contig size:  | 9,968 | 9,968
Mean contig size:  | 286,139 | 289,184
Median contig size:  | 70,931 | 70,856
Contig N50: | 1,812,216 | 1,815,257
E size (G=asm_size): | 3,063,937 | 3,631,802
Contig N50.0 (G=210m): | 3,701,579 | 3,738,961
Contig N50.0 (G=350m): | 1,083,443 | 1,134,487
E size (G=210m): | 4,262,493 | 5,051,241
E size (G=350m): | 2,557,496 | 3,030,744

It seems the 3x setting is catching up to the 1.0 size stats (e.g. similar max), but still not quite there (e.g. still smaller asm by 10-20 Mb.


As for issues brought up in #63 -- it appears better as well. Recall that when low cov settings were used (corMhapSensitivity=high corMinCoverage=2), 

Metric | default settings 1.1-new | default settings 1.1-old | low cov settings 1.1-new | low cov settings 1.1-old
------- | -------------------- | ------------------- | --------------------- | -------------------
input cov | 42.8  | 42.8 | 42.8  | 42.8
cov after cor | 25.8 | 21.16 | 29.1 | 16.2
trim guessed X cov | 27 | 22 | 30 | 93
cov after trim | 25.8 | 21.1 | 28.99 | -
unitig guessed X cov | 27 | 22 | 29 | -

I don't have those "-" data points, but I recall the trend would be something like 16 and >93.

**Comment by JohnUrban** [comment]
Another example example of how 1.0, 1.1-old, and 1.1-new are doing in terms of size stats.
Same data, low cov settings used 
 corMhapSensitivity=high \
 corMinCoverage=2 \
 errorRate=0.035 

Metric | Canu 1.0 | Canu 1.1-old* | Canu 1.1-new
----------| ------------|-----------------|-------------------------
Number contigs: | 2,419 |  3,358 | 861
Assembly size: | 320,253,333 | 245,220,388 | 297,335,484
Max contig size: | 9,453,071  | 1,331,434 | 10,663,865
Min contig size: | 583 | 5,896 | 8,259
Mean contig size:  | 132,391 | 73,026 | 345,337
Median contig size: | 13,331 | 37,367 | 70,476
Contig N50: | 2,040,142 | 131,805 | 1,230,567
E size (G=asm_size): | 2,624,923 | 226,848 | 2,015,629
Contig N50 (G=210m): | 3,328,119 | 170,466 | 2,183,627
Contig N50 (G=350m): | 1,607,852 | 60,763 | 908,649
E size (G=210m): | 4,003,049 | 264,894 | 2,853,896
E size (G=350m): | 2,401,829 | 158,936 | 1,712,338

*Genome size was set to 250m here (but 310m for the others)
Again - it looks better this time around than the old 1.1. The asm size went back towards 300mb, but not as high as 1.0. The max went back up. The min is high as with the other asm. The contig NG50s went back up but not quite as high as 1.0. Interestingly, the new 1.1 has way fewer contigs than the previous 1.1 and 1.0. Again, there are several hundred contigs described in the consensus.out files that are not in contigs.fasta and the asm size is ~7 Mb smaller than if I calculate from those -- and again the "missing" ones seem to be in the unassembled.fasta file.

So it looks like the low-cov tiny contig problem is roughly fixed and, as my last post showed, the Guessed X cov problem for low-cov settings seems to have been fixed.

**Comment by woodoo46** [comment]
Hi, skoren

I have similiar issue, although I am not sure if it is because of our data or the program. We have only about 27X coverage on a 3G genome. I just want to clarify something here:
1) we used the default parameter suggested in Canu document:
corMhapSensitivity=high corMinCoverage=2 errorRate=0.035 minOverlapLength=499 corMaxEvidenceErate=0.3

Is the corMaxEvidenceErate correct?

2) we checked out the code on March 11th. Should we recheck the code and rerun?

Thanks.

**Comment by skoren** [comment]
1) I don't think it's related. The 27X is definitely going to limit your assembly, what is your N50 now? On human/mammalian genomes it is typically in the 500kb-1mb range with that much coverage but that will be limited by your genome complexity and heterozygosity.

2) The code on march 11th was between releases and could have had bugs so I'd get the latest release and try running.

**Comment by skoren** [comment]
This is fixed in v1.2. @woodoo46 can you open your issue as a new ticket so we can track it separately.
